### PR TITLE
Improved progress bar

### PR DESCRIPTION
**Pull request by bwaresiak**
Task/Issue URL: https://app.asana.com/0/856498667320406/1117573875936431
Tech Design URL:
CC:

**Description**:
Improvements to progress bar:
- It is now moved above Omni Bar.
- Smooth animation on update.
- Gradient animation for 'filled' part of the progress bar.

**Steps to test this PR**:
Regular scenario:
1. Open new tab.
2. Start loading webpage.
3. Look out for any glitches/issues around progress bar.

**Redirect**:
1. Open new tab.
2. Navigate to page that redirects to another domain, e.g. bbc.co.uk
3. Look out for any glitches/issues around progress bar.

**Navigating across tabs**:
1. Open new tab.
2. Navigate to page that loads slowly.
3. Open another tab, load another web page there.
4. Progress bar should reflect actual progres when switching between tabs.

**Clearing data**:
1. Open new tab.
2. Navigate to page that loads slowly.
3. Use fire button to clear data.
4. Progress bar should disappear.

**Gradient animation checks**:
1. Open new tab.
2. Navigate to page that loads slowly.
3. Open privacy protection dashboard, then close it OR put app to the background, then move back to it.
4. If visible, progress bar gradient should be animating.

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

File: DuckDuckGo/ProgressView.swift [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+        guard (progressMask.animationKeys() ?? []).isEmpty,
+            currentProgress > visibleProgress else {
+            return
+        }
+        
+        let progressFrame = calculateProgressMaskRect()
+        visibleProgress = currentProgress
+        
+        let duration: TimeInterval
+        if animated == false {
```
Comment (brindy) [inline DuckDuckGo/ProgressView.swift]: `!animated`

File: DuckDuckGo/ProgressView.swift [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+    
+    func hide(animated: Bool = false) {
+        if animated {
+            CATransaction.begin()
+            let animation = CABasicAnimation(keyPath: "opacity")
+            animation.fromValue = 1
+            animation.toValue = 0
+            animation.duration = 0.4
+            progressMask.add(animation, forKey: Constants.fadeOutAnimationKey)
+            CATransaction.setCompletionBlock {
```
Comment (brindy) [inline DuckDuckGo/ProgressView.swift]: Just reference the function directly?
`CATransaction.setCompletionBlock(stopGradientAnimation)`

File: DuckDuckGo/ProgressView.swift [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+        progressMask.add(animation, forKey: Constants.progressAnimationKey)
+        
+        CATransaction.begin()
+        CATransaction.setDisableActions(true)
+        progressMask.bounds = progressFrame
+        CATransaction.commit()
+    }
+    
+    func animationDidStop(_ anim: CAAnimation, finished flag: Bool) {
+        if visibleProgress > 1 - CGFloat.ulpOfOne {
```
Comment (brindy) [inline DuckDuckGo/ProgressView.swift]: Could define `1 - CGFloat.ulpOfOne` as a constant and reference it from both places?

**Review by brindy — COMMENTED** [review]
looks slick! just a few minor comments

File: DuckDuckGo/ProgressView.swift (hunk 3 above)
Comment (bwaresiak) [reply DuckDuckGo/ProgressView.swift]: Hm, I wonder about that. I'd consider keeping 'ulp' here as it is and I'm leaning towards the idea that only numbers that have "meaning" should be defined (and for me number with meaning here is 1.0, representing 100% progress).

If you still would prefer to have it defined, I have two questions:
- Any idea for the naming? (nearlyOne ? 🤦‍♂ )
- Should it be as a constant here, or more global one as CGFloat category?

Comment (brindy) [reply DuckDuckGo/ProgressView.swift]: In terms of naming to me it's a "completion threshold", so `completionThreshold` might be appropriate?  And probably just be a constant in this class.

Comment (bwaresiak) [reply DuckDuckGo/ProgressView.swift]: Thanks @brindy!

**Review by brindy — APPROVED** [review]
LGTM! looks great

**Comment by bwaresiak** [comment]
@brindy Thank you!

**Pull request merged**